### PR TITLE
docs: add custom branch protection examples

### PR DESCRIPTION
**Pull request by jd**
Fixes MRGFY-2428

Change-Id: If6491382ad44e15eddb56c4a87eb27f4ec76ec5e

### CODE REVIEW

**Review by AlexandreGaubert — DISMISSED** [review]

File: src/content/workflow/custom-branch-protections.mdx [hunk 1]
```diff
@@ -0,0 +1,264 @@
+---
+title: Advanced Branch Protection with Mergify Workflow Automation
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:2]:
```suggestion
title: Advanced Branch Protections with Mergify Workflow Automation
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+## Why Mergify Over Native GitHub Branch Protection?
+
+1. **Advanced Conditions**: Mergify supports a vast array of conditions, which
+   means you can create extremely specific rules.
+
+2. **No Conflicts**: Using Mergify avoids potential conflicts with features
+   like the [Mergify Merge Queue](/merge-queue/).
+
+3. **Streamlined Workflows**: Combine branch protection with other workflow
+   actions for a holistic approach.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:27]:
```suggestion
3. **Streamlined Workflows**: Combine branch protections with other workflow
   actions for a holistic approach.
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+There are two common ways of writing custom branch protection settings using
+Mergify:
+
+- Integrate your custom branch protection rules directly into your [automatic
+  merge](automerge) rules;
+
+- Write custom checks that are posted by Mergify on pull requests like a CI
+  would do. This custom checks can also be used as conditions in the automatic
+  merge conditions.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:37]:
```suggestion
  would do. These custom checks can also be used as conditions in the automatic
  merge conditions.
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        - and:
+          - author = dependabot[bot]
+          - "#approved-reviews-by>=1"
+    actions:
+      merge:
+```
+
+With this, Mergify will automatically trigger the merge of a pull request if
+author is [Dependabot](https://github.com/dependabot) and has one approval, or
+if the PR has a two approvals.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:61]:
```suggestion
if the PR has two approvals.
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+This more flexible than GitHub protected branch which only offers a fixed
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:63]:
```suggestion
This is more flexible than GitHub protected branch which only offers a fixed
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+You could auto-merge pull requests only modifying documentation.
+
+```yaml
+pull_request_rules:
+  - name: Documentation PRs
+    conditions:
+      # At least one modified files in the docs/ directory
+      - files ~= ^docs/
+      # No other files are modifeid#
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:76]:
```suggestion
      # No other files are modified
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+              - or:
+                  - approved-reviews-by = @eng-mgr
+                  - and:
+                      - title ~= ^(revert|fix)
+                      - "#approved-reviews-by >= 1"
+    actions:
+      merge:
+```
+
+This allows pull request to be automatically merged with conditions that are
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:115]:
```suggestion
This allows pull requests to be automatically merged with conditions that are
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        - label = hotfix
+    actions:
+      merge:
+```
+
+## Leveraging Custom Checks
+
+Mergify offers an action named [post check](/workflow/actions/post_check) which
+allows to post custom checks. This allows to implement feedback to developers
+like a CI system would do, allowing to enforce extra rules on pull request
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:140]:
```suggestion
like a CI system would do, allowing to enforce extra rules on pull requests
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+```yaml
+pull_request_rules:
+  - name: PR Approval
+    conditions:
+      - base = main
+    actions:
+      post_check:
+        success_condition:
+          - or:
+            # Regular approva is 2 reviews
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:206]:
```suggestion
            # Regular approval is 2 reviews
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+                # If the PR author is a member of GitHub "devs" team, they need
+                # one approval from a manager unless PR is a revert/fix.
+                - author = @devs
+                - or:
+                  - approved-reviews-by = @eng-mgr
+                  - and:
+                      - title ~= ^(revert|fix)
+                      - "#approved-reviews-by >= 1"
+          # You can default to neutral rather than failure if the rule does not succeed
+          # (Note that if you're using this with GitHub branch protection,
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:225]:
```suggestion
          # (Note that if you're using this with GitHub branch protections,
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+- it received 2 approvals;
+
+- it has the label "hotfix", the author is part of the `eng-mgr` team and it
+  has been approved by somebody else;
+
+- it has the label "hotfix", the author is part of the `devs` team and it has
+  been approved by a member of the `eng-mgr` team;
+
+- it has the label "hotfix", the author is part of the `devs` team and it has
+  been approved by somebody else and it fixes an issue or revert a commit.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:248]:
```suggestion
  been approved by somebody else and it fixes an issue or reverts a commit.
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+If one of those scenario is true, the check `Rule: PR Approval (post_check)`
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:250]:
```suggestion
If one of those scenarios is true, the check `Rule: PR Approval (post_check)`
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+### Documentation PRs
+
+You could auto-merge pull requests only modifying documentation.
+
+```yaml
+pull_request_rules:
+  - name: Documentation PRs
+    conditions:
+      # At least one modified files in the docs/ directory
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:74]:
```suggestion
      # At least one modified file in the docs/ directory
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    actions:
+      merge:
+```
+
+With this, Mergify will automatically trigger the merge of a pull request if
+author is [Dependabot](https://github.com/dependabot) and has one approval, or
+if the PR has two approvals.
+
+This is more flexible than GitHub protected branch which only offers a fixed
+number of approval for any pull request.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:64]:
```suggestion
This is more flexible than GitHub protected branch which only offers a fixed
number of approvals for any pull request.
```

File: src/content/workflow/custom-branch-protections.mdx [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+                  - approved-reviews-by = @eng-mgr
+                  - and:
+                      - title ~= ^(revert|fix)
+                      - "#approved-reviews-by >= 1"
+    actions:
+      merge:
+```
+
+This allows pull requests to be automatically merged with conditions that are
+finer-grained that GitHub generic branch protections.
```
Comment (DouglasBlackwood) [inline src/content/workflow/custom-branch-protections.mdx:116]:
```suggestion
finer-grained than GitHub generic branch protections.
```

**Review by DouglasBlackwood — APPROVED** [review]

File: src/content/workflow/custom-branch-protections.mdx [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+    actions:
+      merge:
+```
+
+This allows pull requests to be automatically merged with conditions that are
+finer-grained than GitHub generic branch protections.
+
+### Scheduling Merges
+
+You can protect merges to happen only during office hours. For urgent fixes,
```
Comment (lecrepont01) [inline src/content/workflow/custom-branch-protections.mdx:120]:
```suggestion
You can schedule merges to happen only during office hours. For urgent fixes,
```
or so they can happen maybe

File: src/content/workflow/custom-branch-protections.mdx [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+```
+
+The result of this check will be visible in GitHub user interface, near the
+merge button.
+
+<Screenshot src={postCheckActions} alt="Mergify Post Check" />
+
+Checks posted by Mergify using this action are usable for any other conditions.
+Taking the, you could use the posted check as a condition to merge
+automatically your pull request:
```
Comment (lecrepont01) [inline src/content/workflow/custom-branch-protections.mdx:182]:
```suggestion
You could use the posted check as a condition to automatically merge your pull request:
```

**Review by DouglasBlackwood — APPROVED** [review]

**Review by Greesb — APPROVED** [review]

**Pull request merged**